src/compiler/checker.ts

```diff
@@ -13549,6 +13567,57 @@ namespace ts {
                 return caseType.flags & TypeFlags.Never ? defaultType : getUnionType([caseType, defaultType]);
             }
 
+            function narrowBySwitchOnTypeOf(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
+                const switchWitnesses = getSwitchClauseTypeOfWitnesses(switchStatement);
+                if (!switchWitnesses.length) {
+                    return type;
+                }
+                const clauseWitnesses = switchWitnesses.slice(clauseStart, clauseEnd);
+                //  Equal start and end denotes implicit fallthrough; undefined marks explicit default clause
+                const hasDefaultClause = clauseStart === clauseEnd || contains(clauseWitnesses, /*explicitDefaultStatement*/ undefined);
+                const switchFacts = getFactsFromTypeofSwitch(clauseStart, clauseEnd, switchWitnesses, hasDefaultClause);
+                // The implied type is the raw type suggested by a
+                // value being caught in this clause.
+                // - If there is a default the implied type is not used.
+                // - Otherwise, take the union of the types in the
+                // clause. We narrow the union using facts to remove
+                // types that appear multiple types and are
+                // unreachable.
+                // Example:
+                //
+                // switch (typeof x) {
+                //     case 'number':
+                //     case 'string': break;
+                //     default: break;
+                //     case 'number':
+                //     case 'boolean': break
+                // }
+                //
+                // The implied type of the first clause number | string.
+                // The implied type of the second clause is string (but this doesn't get used).

```

Comment:
Is this comment correct? It seems a little off. Also could use a test which validates fallthrough behaviors, ie,
```ts
let x: string | number | boolean | object;
switch (typeof x) {
    case 'number':
        assertNumber(x)
    case 'string':
        assertStringOrNumber(x)
        break;
    default:
        assertObject(x);
    case 'number':
    case 'boolean':
        assertBooleanOrObject(x);
        break;
 }
```

Reply:
Yes, the comment is wrong. The line for the second clause should read something like:

_"The implied type of the second clause is never, but this does not get just because it includes a default case"_